### PR TITLE
Avoid IEnumerable allocations when flusing data to sqlite.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/23425

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @Therzok @sharwell

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
This saves around 3m allocations, and 300MB of garbage.

**Comment by jasonmalinowski** [comment]
Tagging @Pilchie for ask mode approval.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs [hunk 1]
```diff
@@ -129,8 +129,13 @@ async Task<(Task previousTask, TaskCompletionSource<int> taskCompletionSource)>
                 // performed by FlushAllPendingWritesAsync.
                 using (await _writeQueueGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
                 {
-                    // Get the writes we need to process.
-                    writesToProcess.AddRange(keyToWriteActions[key]);
+                    // Get the writes we need to process. 
+                    // Note: explicitly foreach so we operate on the struct enumerator for
+                    // MultiDictionary.ValueSet.
+                    foreach (var action in keyToWriteActions[key])
```
Comment (Therzok) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs]: Can we extend the capacity with `keyToWriteActions.Count` first?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs]: Let me check!

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs]: Done.

**Comment by Therzok** [comment]
Much better!

**Comment by CyrusNajmabadi** [comment]
> Much better!

Thanks!  Keep the memory reports coming.  They're invaluable!

**Comment by Pilchie** [comment]
Can someone please fill out the ask mode template?

**Comment by CyrusNajmabadi** [comment]
> Can someone please fill out the ask mode template?

@sharwell can you help here?

**Comment by sharwell** [comment]
@CyrusNajmabadi Yes I'll get it added later tonight

**Comment by sharwell** [comment]
### Customer scenario

Large solutions have observable performance overhead relative to small projects, and measurements showed this as relatively high value:risk improvement.

### Bugs this fixes

Fixes #23425

### Workarounds, if any

None needed.

### Risk

Low, this is the standard way of approaching this type of issue.

### Performance impact

Improves performance by avoiding unnecessary allocations on a measured hot path.

### Is this a regression from a previous update?

No.

### Root cause analysis

The original implementation was simple; no change was needed prior to measuring it. A comment was added to avoid regressions now that we know about this path.

### How was the bug found?

Profiling.

### Test documentation updated?

N/A

**Comment by sharwell** [comment]
@Pilchie the template is ready now

**Comment by Pilchie** [comment]
Thanks - approved.

**Pull request merged**